akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala

```diff
@@ -12,56 +12,96 @@ import scala.concurrent.duration.Duration
 import scala.concurrent.duration.FiniteDuration
 
 private[akka] object ReceiveTimeout {
-  final val emptyReceiveTimeoutData: (Duration, Cancellable) = (Duration.Undefined, ActorCell.emptyCancellable)
+  import ActorCell.emptyCancellable
+
+  private final val empty: ReceiveTimeoutValue =
+    ReceiveTimeoutValue(Duration.Undefined, emptyCancellable, Duration.Undefined)
+
+  final case class ReceiveTimeoutValue(current: Duration, task: Cancellable, previous: Duration) {
+
+    def :+(latest: Duration): ReceiveTimeoutValue = latest match {
+      case null => this // user input check
+      case _    =>
+        val prev = current
+        copy(previous = prev, current = latest)
+    }
+
+    /** Returns true if `current` is different from `previous` duration. */
+    def changed: Boolean = current ne previous
+
+    def clear(): ReceiveTimeoutValue = copy(task = emptyCancellable)
+
+    /* read-only functions */
+    def nonEmpty: Boolean = this ne empty
+    def taskIsEmpty: Boolean = task eq emptyCancellable
+    def taskNonEmpty: Boolean = !taskIsEmpty
+  }
+
+  object ReceiveTimeoutValue {
+    def apply(): ReceiveTimeoutValue = empty
+
+    def apply(duration: FiniteDuration, task: Cancellable): ReceiveTimeoutValue =
+      ReceiveTimeoutValue(duration, task, Duration.Undefined)
+  }
 }
 
 private[akka] trait ReceiveTimeout { this: ActorCell =>
 
   import ReceiveTimeout._
-  import ActorCell._
 
-  private var receiveTimeoutData: (Duration, Cancellable) = emptyReceiveTimeoutData
+  private var _receiveTimeout: ReceiveTimeoutValue = ReceiveTimeoutValue()
 
-  final def receiveTimeout: Duration = receiveTimeoutData._1
+  final def receiveTimeout: Duration = _receiveTimeout.current
 
-  final def setReceiveTimeout(timeout: Duration): Unit = receiveTimeoutData = receiveTimeoutData.copy(_1 = timeout)
+  final def setReceiveTimeout(timeout: Duration): Unit = _receiveTimeout :+= timeout
 
-  protected def checkReceiveTimeoutIfNeeded(message: Any): Unit =
-    if (hasTimeoutData)
-      checkReceiveTimeout(!message.isInstanceOf[NotInfluenceReceiveTimeout])
+  protected def rescheduleOrCancelReceiveTimeoutIfNeeded(message: Any): Unit =
+    if (_receiveTimeout.nonEmpty)
+      rescheduleOrCancelReceiveTimeout(shouldInfluenceReceiveTimeout(message))
```

Comment:
These are not really/necessarily rescheduling since there is no check, they cancel regardless of whether there is an existing task and proceed to schedule. But I kept the `reschedule` because they _could_ be a reschedule.